### ISSUE TITLE
Guards, aka when-clauses, are under-specified for grouped matches

### ISSUE THREAD

**Issue by abelbraaksma** [Microsoft/visualfsharp#4666]
Looking at this code, can you guess what will happen?

```f#
    let (|A|B|) x =
        if x = 1 then A
        else B

    let foo() =
        let test = Some 1
        match test with
        | Some(A as a)
        | Some(B as a) when a > 12 -> a
        | Some(B as b) -> b
        | None -> 42
```

Perhaps it is obvious, but it wasn't for me (and of course, it is a slimmed-down version of a more elaborate real-world situation where it was less trivial).

But if you guessed correctly that the when-clause *does* apply to the set of test before it you have probably read the specification in great detail ;).

I understand now that this is a disjunctive pattern, which is not the same as a pattern rule, they just happen to look precisely the same.

In cases like the above, we get a warning "Incomplete pattern matches on this expression". This is good, but often it doesn't cite what it's missing. I understand that can be non-trivial, so I suggest something along these lines:

* If the warning can be improved, then perhaps we should try to in cases like these
* It's quite easy to make the above mistake and it is difficult to spot, perhaps we could create a low-level warning about the specific case of disjunctive patterns with when-guards
* Improve the documentation online (nothing about this is mentioned around the guards-section) and perhaps also add one or two examples to the F# Spec (section 7.13). I actually noticed a lot of errors in the examples in 7.x sections, it could use some attention ;).
* other ideas to improve in this area?

**Comment by dsyme** [comment]
Yes, the when clause applies to both.  The spec and docs should be improved.  

This probably isn't the right place to track this - perhaps just submit a PR to docs.microsoft.com for now?  Or open an issue in that repo.

**Comment by abelbraaksma** [comment]
@dsyme: thanks, I'll do that. Last time I did that I clicked the Edit button on docs.microsoft.com and @cartermp told me it directed me to the wrong location / github repro, leading to him having to make the changes again on the "other" docs repo. If this is fixed, 

I.e., if you go that route, your fixes end up like this: https://github.com/MicrosoftDocs/visualfsharpdocs/pull/308.

**Comment by cartermp** [comment]
@abelbraaksma I think you might have clicked something else. The edit button in the link @dsyme provides (and anything on that site) takes you to github/dotnet/docs, where things are very much "live" and where all conceptual F# documentation lives. The Visual F# Docs on MSDN is only a frozen-in-time FSharp.Core API reference.

**Comment by abelbraaksma** [comment]
@cartermp, thanks, I see it now. It is rather confusing. For instance, when Google brings me here: https://msdn.microsoft.com/en-us/visualfsharpdocs/conceptual/casting-and-conversions-%5bfsharp%5d, then when I click (most of the?) links in the table of contents, it brings me to Docs (different layout). They appear to be HTTP redirects or rewrites. But clicking "Contribute" on that MSDN page, it doesn't redirect and brings me to MicrosoftDocs github.

Seems like I was just unlucky with that particular page still living in the realm of MSDN.

The way I see that is that there are some pages still on MSDN, but most have moved to Docs? Either way, good to know and I'll be careful next time.
